core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala

```diff
@@ -133,20 +166,173 @@ private[op] class OpMultiClassificationEvaluator
+    }
+  }
+
+  object SearchHelper extends Serializable{
+
+    def findThreshold(arr: IndexedSeq[Double], element: Double): Double = {
+      if (element > arr.last) arr.last
+      else if (element < arr(0)) 0.0
+      else {
+        val insertionPoint = new SearchImpl(arr).search(element).insertionPoint
+        val insertionPointValue = arr(insertionPoint)
+        if (element == insertionPointValue) insertionPointValue
+        else arr(insertionPoint-1)
+      }
+    }
+  }
+
+/**
+ * function to calculate confusion matrix for TopK most occurring labels by confidence threshold
+ *
+ * @param data RDD of (label, prediction, prediction probability vector)
+ * @return a MulticlassConfMatrixMetricsByThreshold object
+*/
+
+  def calculateConfMatrixMetricsByThreshold(
+    data: RDD[(Double, Double, Array[Double])]): MulticlassConfMatrixMetricsByThreshold = {
+
+    val labelCountsRDD = data.map { case (label, _, _) => (label, 1L) }.reduceByKey(_ + _)
+    val cmClasses = labelCountsRDD.sortBy(-_._2).map(_._1).take($(confMatrixNumClasses)).toSeq
+    val cmClassesSet = cmClasses.toSet
+
+    val dataTopNLabels = data.filter { case (label, prediction, _) =>
+      cmClassesSet.contains(label) && cmClassesSet.contains(prediction)
+    }
+
+    val sortedThresholds = $(confMatrixThresholds).sorted.toIndexedSeq
+
+    // reduce data to a coarser RDD (with size N * N * thresholds at most) for further aggregation
+    val labelPredictionConfidenceCountRDD = dataTopNLabels.map{
+      case (label, prediction, proba) => {
+        ( (label, prediction, SearchHelper.findThreshold(sortedThresholds, proba.max)), 1L )
+      }
+    }.reduceByKey(_ + _)
+
+    labelPredictionConfidenceCountRDD.cache()
+
+    val cmByThreshold = sortedThresholds.map( threshold => {
+      val filteredRDD = labelPredictionConfidenceCountRDD.filter {
+        case ((_, _, confidence), _) => confidence >= threshold
+      }
+      constructConfusionMatrix(filteredRDD, cmClasses)
+    })
+
+    MulticlassConfMatrixMetricsByThreshold(
+      confMatrixNumClasses = $(confMatrixNumClasses),
+      confMatrixClassIndices = cmClasses,
+      confMatrixThresholds = $(confMatrixThresholds),
+      confMatrices = cmByThreshold
+    )
+  }
+
+/**
+ * function to calculate the mostly frequently mis-classified classes for each label/prediction category
+ *
+ * @param data RDD of (label, prediction)
+ * @return a MisClassificationMetrics class
+ */
+  def calculateMisClassificationMetrics(data: RDD[(Double, Double)]): MisClassificationMetrics = {
+
+    val labelPredictionCountRDD = data.map {
+      case (label, prediction) => ((label, prediction), 1L) }
+      .reduceByKey(_ + _)
+
+    val misClassificationsByLabel = labelPredictionCountRDD.map {
+      case ((label, prediction), count) => (label, (prediction, count))
+    }.groupByKey()
+      .map { case (label, predictionCountsIter) => {
+        val misClassificationCtMap = predictionCountsIter.toSeq
```

Comment:
Here and below - instead of iterating through predictionCountsIter twice for cases where `pred != label` and `pred == label`, better use `predictionCountsIter.partition(...)`

Reply:
@tovbinm  followed gera's suggestion and changed it to  `.collect { case (pred, count) if pred == label => count }`

Reply:
ok.